File: src/lib/ipc/commands.ts
import { invoke } from '@tauri-apps/api/core';
import type { AppConfig } from './types';

// Typed wrappers around the Rust `#[tauri::command]` handlers in
// src-tauri/src/commands.rs. Each call returns a Promise and rejects
// with the string body emitted by `Err(String)` on the Rust side.

export const commands = {
  loadConfig: (): Promise<AppConfig> => invoke<AppConfig>('load_config'),

  saveConfig: (config: AppConfig): Promise<void> => invoke<void>('save_config', { config }),

  resetConfig: (): Promise<AppConfig> => invoke<AppConfig>('reset_config'),

  getAppVersion: (): Promise<string> => invoke<string>('get_app_version'),

  quitApp: (): Promise<void> => invoke<void>('quit_app'),

  isFullscreen: (): Promise<boolean> => invoke<boolean>('is_fullscreen'),

  /** Toggle the main window's fullscreen state. Resolves with the new state. */
  toggleFullscreen: (): Promise<boolean> => invoke<boolean>('toggle_fullscreen'),
};

// ── Translation IPC (M3) ──────────────────────────────────────────────────────

/** Pixel-coordinate region within the video frame (top-left origin). */
export type Region = { x: number; y: number; width: number; height: number };

/**
 * Live status of the offline translation model.
 * Mirrors `ModelStatus` in src-tauri/src/translation/types.rs.
 *
 * The Rust enum uses `#[serde(tag = "type", rename_all = "snake_case")]` so
 * each variant arrives as `{ type: 'not_installed' }`, `{ type: 'downloading',
 * bytes: number, total: number }`, etc.
 */
export type ModelStatus =
  | { type: 'not_installed' }
  | { type: 'downloading'; bytes: number; total: number }
  | { type: 'ready' }
  | { type: 'failed'; message: string };

/** Result of a single OCR-translate cycle. */
export type OcrTranslateResult = {
  en: string;
  th: string;
  latency_ms: number;
  cache_hit: boolean;
  duplicate: boolean;
};

/**
 * OCR a captured JPEG frame and translate the text to Thai.
 *
 * `jpegBytes` — JPEG-encoded frame crop from the hidden canvas.
 * `region`    — pixel region used to crop inside the JPEG (or null to use the
 *               full frame).  The crop is re-applied server-side by Apple Vision.
 *
 * Resolves with `null` when the frame contains no text or is a near-duplicate
 * (the frontend should keep the previous translation visible).
 *
 * Note: Tauri 2 serialises `Vec<u8>` as a JSON array of numbers.
 * `Array.from(Uint8Array)` produces exactly that format.
 */
export const ocrTranslate = (
  jpegBytes: Uint8Array,
  region: Region | null,
): Promise<OcrTranslateResult | null> =>
  invoke<OcrTranslateResult | null>('ocr_translate', {
    jpegBytes: Array.from(jpegBytes),
    region,
  });

/** Query the current model status without downloading anything. */
export const getTranslationModelStatus = (): Promise<ModelStatus> =>
  invoke<ModelStatus>('get_translation_model_status');

/**
 * Start a first-run model download (~900 MB).
 *
 * Pass an optional `modelId` to download a specific catalogue entry.
 * When omitted, downloads the currently active model (legacy behaviour).
 *
 * While downloading, the Rust side emits `model-download-progress` events.
 * The payload is `{ model_id: string, status: ModelStatus }` when a model_id
 * is provided, or a bare `ModelStatus` for the legacy path.
 */
export const downloadTranslationModel = (modelId?: string): Promise<void> =>
  invoke<void>('download_translation_model', { modelId: modelId ?? null });

// ── Model picker commands (M5.5) ──────────────────────────────────────────────

/** Per-model catalogue entry returned by `listTranslationModels`. */
export type ModelInfo = {
  id: string;
  display_name: string;
  description: string;
  /** Estimated download size in bytes. */
  size_bytes: number;
  /** True when model files are present and verified on disk. */
  installed: boolean;
  /** True when this is the currently active (loaded) model. */
  is_active: boolean;
  /** Actual on-disk size in bytes (null when not installed). */
  installed_size_bytes: number | null;
};

/** Return metadata for all models in the catalogue. */
export const listTranslationModels = (): Promise<ModelInfo[]> =>
  invoke<ModelInfo[]>('list_translation_models');

/**
 * Delete a model's files from disk.
 *
 * Rejects with an error string when the model is currently active and loaded.
 */
export const deleteTranslationModel = (modelId: string): Promise<void> =>
  invoke<void>('delete_translation_model', { modelId });

/**
 * Switch the active translation model.
 *
 * Unloads the current translator and loads the new one from disk.
 * Rejects with an error string if the model is not installed.
 */
export const setActiveTranslationModel = (modelId: string): Promise<void> =>
  invoke<void>('set_active_translation_model', { modelId });

File: src-tauri/src/commands.rs
use std::path::PathBuf;
use std::sync::Arc;

use tauri::{AppHandle, Manager, Runtime};
use tokio::sync::Mutex;

use crate::config::{self, AppConfig};
use crate::translation::{
    engine::{ModelStatusHandle, TranslationEngine},
    EngineError, ModelStatus, OcrTranslateResult, Region,
};

/// Shared state type for the translation engine.
pub type TranslationEngineState = Arc<Mutex<TranslationEngine>>;

const MAIN_WINDOW_LABEL: &str = "main";

fn main_window<R: Runtime>(app: &AppHandle<R>) -> Result<tauri::WebviewWindow<R>, String> {
    app.get_webview_window(MAIN_WINDOW_LABEL)
        .ok_or_else(|| format!("window {MAIN_WINDOW_LABEL} not found"))
}

fn config_path() -> Result<PathBuf, String> {
    config::default_config_path().map_err(|e| e.to_string())
}

#[tauri::command]
pub fn load_config() -> Result<AppConfig, String> {
    let path = config_path()?;
    let cfg = config::load(&path).map_err(|e| e.to_string())?;
    log::info!("config loaded from {}", path.display());
    Ok(cfg)
}

#[tauri::command]
pub fn save_config(config: AppConfig) -> Result<(), String> {
    let path = config_path()?;
    config::save(&config, &path).map_err(|e| e.to_string())?;
    log::info!("config saved to {}", path.display());
    Ok(())
}

#[tauri::command]
pub fn reset_config() -> Result<AppConfig, String> {
    let path = config_path()?;
    let cfg = config::reset(&path).map_err(|e| e.to_string())?;
    log::info!("config reset: {}", path.display());
    Ok(cfg)
}

#[tauri::command]
pub fn get_app_version() -> String {
    env!("CARGO_PKG_VERSION").to_string()
}

#[tauri::command]
pub fn quit_app<R: Runtime>(app: AppHandle<R>) {
    log::info!("quit requested");
    app.exit(0);
}

#[tauri::command]
pub fn is_fullscreen<R: Runtime>(app: AppHandle<R>) -> Result<bool, String> {
    let window = main_window(&app)?;
    window.is_fullscreen().map_err(|e| e.to_string())
}

#[tauri::command]
pub fn toggle_fullscreen<R: Runtime>(app: AppHandle<R>) -> Result<bool, String> {
    let window = main_window(&app)?;
    let current = window.is_fullscreen().map_err(|e| e.to_string())?;
    let next = !current;
    window.set_fullscreen(next).map_err(|e| e.to_string())?;
    log::info!("fullscreen toggled: {current} -> {next}");
    Ok(next)
}

// ── Translation commands (M3) ─────────────────────────────────────────────────

/// OCR a captured JPEG frame and translate the recognised English text to Thai.
///
/// `jpeg_bytes` — raw JPEG bytes (Tauri serialises `Vec<u8>` as a JSON array
/// of numbers, which matches `Array.from(Uint8Array)` on the frontend).
///
/// Returns `None` when the frame contains no text or is a near-duplicate of
/// the previous frame (the frontend should keep the last translation visible).
#[tauri::command]
pub async fn ocr_translate(
    state: tauri::State<'_, TranslationEngineState>,
    jpeg_bytes: Vec<u8>,
    region: Option<Region>,
) -> Result<Option<OcrTranslateResult>, String> {
    let mut engine = state.lock().await;
    engine
        .ocr_translate(jpeg_bytes, region)
        .await
        .map_err(|e: EngineError| e.to_string())
}

/// Return the current model status without downloading anything.
///
/// Reads from the `ModelStatusHandle` (an `Arc<RwLock<ModelStatus>>`) that is
/// stored as separate Tauri state.  This never acquires the engine mutex, so
/// it cannot block for 5 minutes during a large download (carry-over B).
#[tauri::command]
pub fn get_translation_model_status(
    status: tauri::State<'_, ModelStatusHandle>,
) -> Result<ModelStatus, String> {
    Ok(status.read().unwrap().clone())
}

/// Download the NLLB-200 model files (first-run, ~900 MB) and load the
/// translator into memory.
///
/// Emits `model-download-progress` events with `ModelStatus::Downloading`
/// payloads while the download is in progress.  The final event payload is
/// `ModelStatus::Ready` on success.
#[tauri::command]
pub async fn download_translation_model(
    app: AppHandle,
    state: tauri::State<'_, TranslationEngineState>,
    status: tauri::State<'_, ModelStatusHandle>,
) -> Result<(), String> {
    let mut engine = state.lock().await;
    engine
        .ensure_ready(&app, &status.inner().clone())
        .await
        .map_err(|e: EngineError| e.to_string())
}
